### PR TITLE
Pathfinding

### PR DESCRIPTION
**Pull request by NikaRodin**
I implemented Dijkstra as the path-finding algorithm. This solves #5. 

Additionally, I created some mock classes and added jest tests to test several edge cases.

The important solutions will be documented in the wiki after review.

### CODE REVIEW

File: client/src/utils/Constants.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export class Constants {
+    static infinity: number = Number.MAX_SAFE_INTEGER;
```
Comment (JakovTomasic) [inline client/src/utils/Constants.ts:2]: You should probably use Typescript's `Infinity` directly. Test if it works and remove this file.

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 2]
```diff
@@ -1,19 +1,92 @@
...
+                
+                if (currentDistance + neighbourDistance < bestDistances.get(neighbourId)
+                    || !bestDistances.has(neighbourId)) {
+
+                    bestDistances.set(neighbourId, currentDistance + neighbourDistance);
+                    unvisitedNodes.add({nodeId: neighbourId, distance: bestDistances.get(neighbourId)})
+                    parents.set(neighbourId, currentNodeId);
+                }
+            }
+        }
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:89]: This algorithm can be refactored to match more traditional implementation. Follow the individual comments for that.

You should also mention somewhere in the code that this is Dijkstra's algorithm. You can write a comment or create a function with a related name. Both solutions are fine in this case.

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 3]
```diff
@@ -1,19 +1,92 @@
...
+        });
+
+        let currentDistance: number;
+        let currentNodeId: number;
+        let allVisited: boolean;
+
+        bestDistances.set(startNodeId, 0);
+        unvisitedNodes.add({nodeId: startNodeId, distance: 0});
+
+        while (true) {
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:48]: this should be `while(unvisitedNodes not empty)` (but written with the correct syntax 😄). That will eliminate the need for `allVisited` variable which is obsolete.

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 4]
```diff
@@ -1,19 +1,92 @@
...
+            unvisitedNodes.shift();
+            neighbours = this.graph.getNeighbours(currentNodeId);
+
+            for (let element of neighbours) {
+                let neighbourId = element.neighbour.id;
+                let neighbourDistance = element.distance;
+                
+                if (currentDistance + neighbourDistance < bestDistances.get(neighbourId)
+                    || !bestDistances.has(neighbourId)) {
+
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:83]: You should first check if bestDistances has this negithbour and then get it's value. This will result in unexpected behaviour if bestDistances does not have this neighbour.

The code should be as follows:
```ts
if (!bestDistances.has(neighbourId) || currentDistance + neighbourDistance < bestDistances.get(neighbourId)) {
```

If bestDistances does not have the neighbour it will return true and stop checking the rest of the if condition.

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 5]
```diff
@@ -1,19 +1,92 @@
...
     findShortestPath(startNodeId: number, endNodeFilter: MapNodeFilter): NavigationDirections {
         let path = this.findPathToNearestNode(startNodeId, endNodeFilter);
         let navNodes = this.convertPathToNavigationNodes(path);
         return this.splitNavigationPathIntoSteps(navNodes);
     }
 
-    private findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
-        throw new Error("Method not implemented.");
+    /*private*/ findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:23]: Remove this random private comment. This is now public. Own it 😆

File: client/src/logic/mock/MapNodeFilterMock.ts
Comment (JakovTomasic) [inline client/src/logic/mock/MapNodeFilterMock.ts:1]: Nice job implementing the filter for single destination nodes. This is not mock 😆. This is how the algorithm will find a node when we know its id. Rename this class to something like MapNodeFileterById (just an example) and remove the file somewhere else.

File: client/tests/MapNavigatorImpl.test.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    expect(() => mapNavigator.findPathToNearestNode(startNodeId, mapNodeFilter)).toThrowError(errorMessage);
+  });
+
+  test('should throw start location error', () => {
+    var startNodeId = 10;
+    var mapNodeFilter = new MapNodeFilterMock(10);
+    var errorMessage = "Start location not valid.";
+
+    expect(() => mapNavigator.findPathToNearestNode(startNodeId, mapNodeFilter)).toThrowError(errorMessage);
+  });
```
Comment (JakovTomasic) [inline client/tests/MapNavigatorImpl.test.ts:117]: Feel free to answer this discussion. Why are you creating separate test cases for hallway nodes? The path-finder implementation doesn't know about hallway vs endNode. It only knows about the MapNode type. Thus, it does not make sense to test it separately.

But as I can see, the test cases for hallways are completely different than for the nodes. If that is the case you should not delete them, but just replace the hallways with EndNodes in the mock data.

Am I missing something? Why do you feel this should be tested separately? Is there some part of the implementation that could break if multiple classes are used?

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 7]
```diff
@@ -1,19 +1,92 @@
...
+
+        while (true) {
+            currentDistance = Constants.infinity;
+            allVisited = true;
+
+            if (unvisitedNodes.length > 0) {
+                currentNodeId = unvisitedNodes.head().nodeId;
+                currentDistance = unvisitedNodes.head().distance;
+                allVisited = false;
+            }
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:56]: After conforming the while to stop when the set is empty, this will not be needed. Just read the head (the first two lines in if are good, but if condition should be removed because it will always be true.

File: client/src/logic/impl/MapNavigatorImpl.ts [hunk 8]
```diff
@@ -1,19 +1,92 @@
...
+
+                    path.push(this.graph.getNode(parentId));
+                    currentNodeId = parentId;
+                }
+
+                return path.reverse();
+
+            } else if (allVisited) {
+                throw new Error("Destination cannot be reached.");
+            }
```
Comment (JakovTomasic) [inline client/src/logic/impl/MapNavigatorImpl.ts:72]: This is the end condition. It returns the result, but it is a bit confusing because it is written in the middle of the 
algorithm code. This code should be moved below the while.

You probably added this code for early stopping before all nodes are visited. You can still achieve that by calling `break` if the filter is satisfied.

Bonus tip - the path recreation algorithm (this little while-loop) is not an essential part of dijkstra and should be extracted into a function.

**Review by JakovTomasic — CHANGES_REQUESTED** [review]
Well done writing the algorithm, tests, and separating types into more reasonable folders.

Also, you have chosen the right location for the tests and the right naming convention for the test file (whatItTests.test.ts). You can document that in the documentation for writing the code.

I requested some code quality changes.

File: client/src/logic/impl/MapNavigatorImpl.ts (hunk 4 above)
Comment (NikaRodin) [reply client/src/logic/impl/MapNavigatorImpl.ts:83]: Actually it doesn't result in an unexpected behaviour. It just returns false.
I wrote it this way because it seemed easier to read, but I will reverse them now.

File: client/tests/MapNavigatorImpl.test.ts (hunk 6 above)
Comment (NikaRodin) [reply client/tests/MapNavigatorImpl.test.ts:117]: You aren't missing anything. There was no need to test the Hallway nodes separately, I just wanted to test a different graph. I will change the misleading name, but do I have to replace the mock data with EndNodes? You said it yourself it doesn't make a difference.

File: client/src/utils/Constants.ts (hunk 1 above)
Comment (NikaRodin) [reply client/src/utils/Constants.ts:2]: Cool! I didn't even know this existed haha

File: client/src/logic/impl/MapNavigatorImpl.ts (hunk 2 above)
Comment (JakovTomasic) [reply client/src/logic/impl/MapNavigatorImpl.ts]: Nice job refactoring. The code is now clean.

File: client/src/logic/impl/MapNavigatorImpl.ts (hunk 4 above)
Comment (JakovTomasic) [reply client/src/logic/impl/MapNavigatorImpl.ts]: It works because of javascript (and typescript) specifics. If bestDistances map doesn't have a key neighbourId it doesn't know what to return (probably 0 or some default in this case). Then you compare that "null" result with numbers. In some other languages that would crash the execution.

Now after the fix, you first checked if the id is in the map and only after that do you read its value.

If you don't understand contact me informally. I'll explain it quickly.

**Review by JakovTomasic — APPROVED** [review]
Everything is fixed now. The code is very good. We just need to resolve the conflicts and you may merge the branch.
